flink-streaming-java/src/test/java/org/apache/flink/streaming/graph/StreamingJobGraphGeneratorNodeHashTest.java

```diff
@@ -422,6 +425,43 @@ public void testManualHashAssignmentForStartNodeInInChain() throws Exception {
 		env.getStreamGraph().getJobGraph();
 	}
 
+	@Test
+	public void testUserProvidedHashing() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironment();
+
+		List<String> userHashes = Arrays.asList("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb");
+
+		env.addSource(new NoOpSourceFunction(), "src").provideAdditionalNodeHash(userHashes.get(0))
+				.map(new NoOpMapFunction())
+				.filter(new NoOpFilterFunction())
+				.keyBy(new NoOpKeySelector())
+				.reduce(new NoOpReduceFunction()).name("reduce").provideAdditionalNodeHash(userHashes.get(1));
+
+		StreamGraph streamGraph = env.getStreamGraph();
+		int idx = 1;
+		for (JobVertex jobVertex : streamGraph.getJobGraph().getVertices()) {
+			Assert.assertEquals(jobVertex.getIdAlternatives().get(1).toString(), userHashes.get(idx));
+			--idx;
+		}
+	}
+
+	@Test
+	public void testUserProvidedHashingOnChainNotSupported() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironment();
+
+		env.addSource(new NoOpSourceFunction(), "src").provideAdditionalNodeHash("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
+				.map(new NoOpMapFunction()).provideAdditionalNodeHash("bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb")
```

Comment:
the fact that this fails the job should probably be documented in the javadocs of ```provideAdditionalNodeHash```.